core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -137,6 +137,68 @@ class AdminClient(val time: Time,
+   *
+   * PurgeDataResult contains either lowWatermark of the partition or exception. We list the possible exception
+   * and their interpretations below:
+   *
+   * - DisconnectException if leader node of the partition is not available. Need retry by user.
+   * - TimeoutException if response is not available within the timeout specified by either Future's timeout or AdminClient's request timeout
+   * - NotEnoughReplicasException if any replica of the partition is offline
+   * - UnknownTopicOrPartitionException if partition doesn't exist
+   * - NotLeaderForPartitionException if broker is not leader of the partition. Need retry by user.
+   * - OffsetOutOfRangeException if the offset is larger than high watermark of this partition
+   *
+   */
+
+  def purgeDataBefore(purgeOffsets: Map[TopicPartition, Long]): Future[Map[TopicPartition, PurgeDataResult]] = {
+    val metadataRequest = new MetadataRequest.Builder(purgeOffsets.keys.map(_.topic()).toSet.toList.asJava)
+    val response = sendAnyNode(ApiKeys.METADATA, metadataRequest).asInstanceOf[MetadataResponse]
+    val errors = response.errors
+    if (!errors.isEmpty)
+      error(s"Metadata request contained errors: $errors")
+
+    val (authorizedPartitions, unauthorizedPartitions) = purgeOffsets.partition{partitionAndOffset =>
+      response.cluster().leaderFor(partitionAndOffset._1) != null}
+
+    val unauthorizedPartitionResults = unauthorizedPartitions.mapValues( _ =>
+      PurgeDataResult(PurgeResponse.INVALID_LOW_WATERMARK, Errors.UNKNOWN_TOPIC_OR_PARTITION.exception()))
+
+    val authorizedPartitionsGroupByLeader = authorizedPartitions.groupBy(partitionAndOffset =>
+      response.cluster().leaderFor(partitionAndOffset._1))
+
+    // prepare requests and generate Future objects
+    val futures = authorizedPartitionsGroupByLeader.map{ case (node, partitionAndOffsets) =>
+      val convertedMap: java.util.Map[TopicPartition, java.lang.Long] = partitionAndOffsets.mapValues(_.asInstanceOf[java.lang.Long]).asJava
+
+      client.send(node, new PurgeRequest.Builder(requestTimeoutMs, convertedMap))
+        .compose(new RequestFutureAdapter[ClientResponse, Map[TopicPartition, PurgeDataResult]]() {
+
+          override def onSuccess(response: ClientResponse, future: RequestFuture[Map[TopicPartition, PurgeDataResult]]) {
+            val purgeResponse = response.responseBody().asInstanceOf[PurgeResponse]
+            val result = purgeResponse.responses().asScala.mapValues(v => PurgeDataResult(v.lowWatermark, v.error.exception())).toMap
+            future.complete(result)
+          }
+
+          override def onFailure(e: RuntimeException, future: RequestFuture[Map[TopicPartition, PurgeDataResult]]) {
+            val result = partitionAndOffsets.mapValues(_ => PurgeDataResult(PurgeResponse.INVALID_LOW_WATERMARK, e))
+            future.complete(result)
+          }
+
+        })
+    }
+
+    // send requests over network to brokers
+    client.poll(0)
```

Comment:
Great point. I have updated the AdminClient to create its own thread to do `client.poll(retryBackoffMs)`. I find it necessary for AdminClient to have its own thread in order to support both syn and async operation.

I have also added `testLogStartOffsetAfterAsyncPurge()` to validate the asyn purge operation.